t/t3001-mpi-personalities.t

```diff
@@ -0,0 +1,44 @@
+#!/bin/sh
+#
+
+test_description="Test that Flux's MPI personalities work"
+
+. `dirname $0`/sharness.sh
+
+# Size the session to one more than the number of cores, minimum of 4
+SIZE=$(test_size_large)
+test_under_flux ${SIZE} wreck
+echo "# $0: flux session size will be ${SIZE}"
+
+# Usage: run_program timeout ntasks nnodes
+run_program() {
+	local timeout=$1
+	local ntasks=$2
+	local nnodes=$3
+	shift 3
+	run_timeout $timeout flux wreckrun $OPTS \
+		    -n${ntasks} -N${nnodes} $*
+}
+
+test_expect_success "intel mpi" '
+  export I_MPI_PMI_LIBRARY="foobar";
+  run_program 5 ${SIZE} ${SIZE} printenv I_MPI_PMI_LIBRARY \
+  	| tee intel.mpi \
+  	&& test "$(uniq intel.mpi)" == "$(flux getattr conf.pmi_library_path)"
+'
+
+OPTS="-o mpi=spectrum"
+test_expect_success "spectrum mpi sets MPI_ROOT" '
+  run_program 5 ${SIZE} ${SIZE} printenv MPI_ROOT \
+  	| tee spectrum-mpi.env \
+    && test "$(uniq spectrum-mpi.env)" == "/opt/ibm/spectrum_mpi"
+'
+
+OPTS="-o mpi=spectrum"
+test_expect_success "spectrum mpi sets the soft stack limit" '
+  run_timeout 5 flux wreckrun -n ${SIZE} -N ${SIZE} $OPTS bash -c "ulimit -s" \
```

Comment:
I gave up trying to get the quoting of `bash -c "ulimit -s"` to play nice with  `run_program`.  If anyone has any pointers or suggestions, I'm open to hearing them. I'm also content leaving it as is.

Reply:
`""ulimit -s""` doesn't work?

Reply:
What you have seems to be working.  What's the problem?

Reply:
Unfortunately not.
`"ulimit -s"` and `""ulimit -s""` both result in the `-s` not being passed to ulimit.
If I put an `echo` in front of the `run_timeout ...` call in `run_program` and then test various quotings, I get:
```
run_timeout 5 flux wreckrun -o mpi=spectrum -n17 -N17 bash -c ulimit -s
run_timeout 5 flux wreckrun -o mpi=spectrum -n17 -N17 bash -c ulimit -s
run_timeout 5 flux wreckrun -o mpi=spectrum -n17 -N17 bash -c 'ulimit -s'
run_timeout 5 flux wreckrun -o mpi=spectrum -n17 -N17 bash -c 'ulimit -s'
```
from these:
```
OPTS="-o mpi=spectrum"
test_expect_success "spectrum mpi sets MPI_ROOT" '                                                                                                                                                 
  run_program 5 ${SIZE} ${SIZE} bash -c "ulimit -s"                                                                                                                                             
'

OPTS="-o mpi=spectrum"
test_expect_success "spectrum mpi sets MPI_ROOT" '                                                                                                                                                 
  run_program 5 ${SIZE} ${SIZE} bash -c ""ulimit -s""                                                                                                                                               
'

CMD="bash -c 'ulimit -s'"
OPTS="-o mpi=spectrum"
test_expect_success "spectrum mpi sets MPI_ROOT" '                                                                                                                                                 
  run_program 5 ${SIZE} ${SIZE} ${CMD}                                                                                                                                               
'

CMD="bash -c 'ulimit -s'"
OPTS="-o mpi=spectrum"
test_expect_success "spectrum mpi sets MPI_ROOT" '                                                                                                                                                 
  run_program 5 ${SIZE} ${SIZE} "${CMD}"
'

```

And none of them succeed. I would have sworn that the last two would work but :man_shrugging:

Reply:
A complex case for me was https://github.com/flux-framework/flux-scalability/blob/master/t/t0001-sched-scale.t#L53 to pass single quotes to awk. Don't know if this will be helpful.

Reply:
?
```
test_expect_success "spectrum mpi sets the soft stack limit" '
    run_timeout 5 flux wreckrun -n ${SIZE} -N ${SIZE} $OPTS bash -c '\''ulimit -s'\'
'

Reply:
```
OPTS="-o mpi=spectrum"
test_expect_success "spectrum mpi sets the soft stack limit" '                                                                                                                                     
  run_program 5 ${SIZE} ${SIZE} bash -c '\''ulimit -s'\'' \                                                                                                                                        
    | tee spectrum-mpi.ulimit \                                                                                                                                                                    
    && test "$(uniq spectrum-mpi.ulimit)" = "10240"                                                                                                                                                
'
```
results in
```
run_timeout 5 flux wreckrun -o mpi=spectrum -n17 -N17 bash -c ulimit -s
```
and the incorrect result.

My bash-fu isn't good enough to know what is really going on but I think the `$*` in `run_program` is what is changing the quoting.

Reply:
Using double quote inside of single-quoted test script works in other sharness tests in the testsuite, see the wreck and aggregator tests, and I think some kvs tests too.

I wonder what is going wrong here.

Reply:
> If I put an echo in front of the run_timeout ... call in run_program and then test various quotings, I get:

But the `echo` doesn't necessarily show the arguments (which may contain whitespace) passed down to the command.

Sometimes I resort to putting a `set -x` before the code and `set +x` after see what is actually being run.

Reply:
Oops, sorry for the noise, somehow my version of this issue wasn't updated with the latest comments.

Reply:
> Sometimes I resort to putting a set -x before the code and set +x after see what is actually being run.

Thanks @grondo. I'll give that a shot.